### ISSUE TITLE
Material sidenav drawer is not displayed correctly in Ionic

### ISSUE THREAD

**Issue by vially** [ionic-team/ionic#17940]
# Bug Report

**Ionic version:**

[x] **4.1.2**

**Current behavior:**

Showing a [material sidenav](https://material.angular.io/components/sidenav/overview) in an Ionic app is not displayed correctly when placed inside `<ion-content />` (see screenshot below).

![sidenav_ion_content](https://user-images.githubusercontent.com/433598/55477876-827ed780-561a-11e9-8464-8514e270f41c.png)

**Expected behavior:**

It should render the same as when placed inside a plain div (see screenshot below):

![sidenav_plain_div](https://user-images.githubusercontent.com/433598/55477870-7e52ba00-561a-11e9-9c30-c038b6f744be.png)

**Steps to reproduce:**

1. Checkout the `material-drawer-content-size` branch from https://github.com/vially/ionic-bugs-demos/tree/material-drawer-content-size
1. `npm install`
1. `ng serve`

**Related code:**

https://github.com/vially/ionic-bugs-demos/tree/material-drawer-content-size

**Other information:**

Might be related to https://github.com/ionic-team/ionic/issues/17920

**Ionic info:**

```
Ionic:

   ionic (Ionic CLI)             : 4.12.0 (/home/vially/.config/yarn/global/node_modules/ionic)
   Ionic Framework               : @ionic/angular 4.1.2
   @angular-devkit/build-angular : 0.13.7
   @angular-devkit/schematics    : 7.2.4
   @angular/cli                  : 7.3.7
   @ionic/angular-toolkit        : 1.4.1

System:

   NodeJS : v10.15.3 (/usr/bin/node)
   npm    : 6.9.0
   OS     : Linux 5.0
```

**Comment by liamdebeasi** [comment]
Hi there,

Thanks for the issue. It looks like this is an issue with `@angular/material`: https://github.com/angular/material2/issues/6743. Note that while the issue has been closed, numerous users at the bottom are still reporting this issue.

You can view the workaround here: https://github.com/angular/material2/issues/6743#issuecomment-435563464.

I am going to close this as this is more of a support request. For further questions with this feel free to post on our [forums](http://forum.ionicframework.com) or our [slack](https://ionicworldwide.herokuapp.com).

Thanks for using Ionic!

**Comment by vially** [comment]
I don't think this issue is necessarily related to angular material issue that you've linked to. I understand that there may be some issues with the material sidenav component itself when used in various responsive and/or interactive scenarios.

But that's not the case here. The demo I've linked to contains the most basic material sidenav example (non-responsive/non-interactive).

And it works fine when placed inside a `<div>` but it breaks down when placed inside an `<ion-content>`.

Could it be that there is something particular about the `<ion-content>` element that actually makes the material component break down?

**Comment by brandyscarney** [comment]
Hey @vially! If I add the following workaround from the issue Liam mentioned to your project I am able to see the sidemenu properly:

```typescript
@ViewChild('sidenavContainer') private sidenavContainer: MatSidenavContainer;

ngAfterViewInit() {
  this.sidenavContainer.autosize = true;
}
```

<img width="400" src="https://user-images.githubusercontent.com/6577830/55490110-32dfe200-5601-11e9-8233-64e8e45a9292.png">

I was also able to randomly get the `div` to act the same way:

<img width="400" src="https://user-images.githubusercontent.com/6577830/55490151-4723df00-5601-11e9-8827-92924f430100.png">

I'm not sure what happened to get the `div` to look the same, but it appears to be a race condition in the resize. The issue does seem to be related to how Material handles resizing the menu contents. 

If you're able to pinpoint the exact problem as an Ionic one, we'd be happy to look into it more, but we simply don't have the resources to dig into how the Material sidenav determines when to resize its contents, which is why this is best for our support channels. Thank you!

**Comment by vially** [comment]
Thanks for taking the time to look more closely into this.

I know the workaround works, but I was more interested in fixing the root problem because it seems to pop up often enough that we thought it might be worth getting to the bottom of it. Also the documentation of the `autoSize` workaround contains this warning: 

> **Use at your own risk!** Enabling this option can cause layout thrashing by measuring the drawers on every change detection cycle.

I agree that it looks like a race condition. We've seen this sizing race conditions in various places in our application with various components from various libraries that it got us thinking that maybe there's something more to this than just bugs in all of these components.

But I think at the root of most of these sizing problems is the fact that in Ionic a standalone component (which is not a page) **has no way to know** when its size has been computed (see discussion in #17920 for more details on this).

In conclusion in Ionic **you cannot build a component that needs to know its size** that works reliably without making it, or one of its parents, aware of when the `ionViewDidEnter` event has been dispatched. This seems to suggest that even *if* the bug is inside the material sidenav component, it might not be possible to be fixed on their side because in Ionic apps there's no hook inside a component to know when its size has been set.

**Comment by adamdbradley** [comment]
Ionic components are lazy loaded, which is a useful feature to not have to request the entire ionic component library to only load the handful of components actually being used. This allows ionic to have faster startup times and request the minimal amount of code required. A side effect of this is that ion-content may not have fully loaded when a parent component has loaded.

However, every ionic component has a `componentOnReady()` method that returns a promise for when it has finished loading. Internally ionic uses this often, for example, `await element.componentOnReady()`. This allows the code to ensure the component has finished loading before continuing. I would recommend to first stay away from doing dom reads to get element sizes since and that'll cause layout thrashing. However, there are still cases you'll need to do this, so I'd recommend using `requestAnimationFrame`, and within it's callback use `await element.componentOnReady()`. Those two combined should help reduce potential layout thrashing issues, and it'll wait for the component to finish loading so you can get an accurate size. And in most cases, by the time the request animation callback fires, the component would have already been ready.
